hardware/kunlunxin_R300p/237377e9/237377e9.patch

```diff
@@ -1,117 +1,25 @@
 From: brianlcy123 <brianlcy123@gmail.com>
-Date: Sun, 24 Nov 2024 19:12:03 +0800
-Subject: [PATCH] [kunlunxin] add patch for mixtral
+Date: Tue, 24 Dec 2024 10:19:56 +0800
+Subject: [PATCH] [kunlunxin] update mixtral patch for py39
 
 ---
- .../megatron/core/dist_checkpointing/strategies/base.py    | 4 ++--
- .../megatron/core/distributed/param_and_grad_buffer.py     | 7 ++++++-
- megatron/megatron/core/transformer/moe/moe_utils.py        | 6 +++---
- megatron/megatron/core/transformer/moe/token_dispatcher.py | 4 ++--
- megatron/megatron/training/checkpointing.py                | 3 ++-
- 5 files changed, 15 insertions(+), 9 deletions(-)
+ megatron/megatron/training/checkpointing.py | 3 ++-
+ 1 file changed, 2 insertions(+), 1 deletion(-)
 
-diff --git a/megatron/megatron/core/dist_checkpointing/strategies/base.py b/megatron/megatron/core/dist_checkpointing/strategies/base.py
-index cc1c83b9..125779a0 100644
---- a/megatron/megatron/core/dist_checkpointing/strategies/base.py
-+++ b/megatron/megatron/core/dist_checkpointing/strategies/base.py
-@@ -6,7 +6,7 @@ from abc import ABC, abstractmethod
- from collections import defaultdict
- from enum import Enum
- from pathlib import Path
--from typing import Any, DefaultDict
-+from typing import Any, DefaultDict, Dict, Tuple
- 
- from ..mapping import CheckpointingException, ShardedStateDict, StateDict
- from .async_utils import AsyncCallsQueue, AsyncRequest
-@@ -20,7 +20,7 @@ class StrategyAction(Enum):
- 
- 
- _import_trigger = None
--default_strategies: DefaultDict[str, dict[tuple, Any]] = defaultdict(dict)
-+default_strategies: DefaultDict[str, Dict[Tuple, Any]] = defaultdict(dict)
- 
- async_calls = AsyncCallsQueue()
- 
-diff --git a/megatron/megatron/core/distributed/param_and_grad_buffer.py b/megatron/megatron/core/distributed/param_and_grad_buffer.py
-index 77ecd7be..c2761c6e 100644
---- a/megatron/megatron/core/distributed/param_and_grad_buffer.py
-+++ b/megatron/megatron/core/distributed/param_and_grad_buffer.py
-@@ -248,6 +248,11 @@ class ParamAndGradBuffer:
-         def _pad(number_to_be_padded: int, divisor: int) -> int:
-             return int(math.ceil(number_to_be_padded / divisor) * divisor)
- 
-+        import math
-+
-+        def _lcm(a, b):
-+            return abs(a * b) // math.gcd(a, b)
-+
-         def _pad_end_of_bucket_if_needed(bucket_end_index: int) -> int:
-             """
-             Pads end index of bucket if using distributed optimizer (to ensure uniform sharding).
-@@ -257,7 +262,7 @@ class ParamAndGradBuffer:
-                 # This also helps cuBLAS pick more efficient algorithms for GEMMs.
-                 # We now ensure that all buckets start at a memory address that is 256-byte
-                 # aligned (128 values since params and grads use >= 16-bit precision).
--                return _pad(bucket_end_index, math.lcm(self.data_parallel_world_size, 128))
-+                return _pad(bucket_end_index, _lcm(self.data_parallel_world_size, 128))
```

Comment:
Why change it to this way? Are these exactly equivalent？

Reply:
First, that change was in the previous PR and is removed in this PR. 
Previously we were using python3.8, math.lcm was not supported yet. 
Second, they are mathmatically equal.